File: Services/WebhookService.cs
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using static SearchStack.PublicApi.Url;

namespace SearchStack.PublicApi;

/// <summary>
/// Webhook subscription management: list the event catalog, create / edit / remove a subscription, and
/// fire a test delivery. The Webhook permissions are carryable on an API key. Secret view/rotate,
/// delivery-log reads and resume stay console-plane.
/// </summary>
public sealed class WebhookService
{
    private const string Path = "webhook";
    private readonly Transport _transport;

    internal WebhookService(Transport transport) => _transport = transport;

    /// <summary>Lists the catalog of subscribable event types, grouped for display.</summary>
    public Task<Response<List<WebhookEventGroupResponse>>> ListEventsAsync(CancellationToken cancellationToken = default) =>
        _transport.GetAsync<List<WebhookEventGroupResponse>>($"{Path}/events", cancellationToken);

    /// <summary>
    /// Creates a webhook subscription. The response includes the signing secret exactly once — store it
    /// now; it cannot be retrieved again from the data plane.
    /// </summary>
    public Task<Response<WebhookSubscriptionWithSecretResponse>> CreateAsync(CreateWebhookSubscriptionRequest request, CancellationToken cancellationToken = default) =>
        _transport.PostAsync<WebhookSubscriptionWithSecretResponse>(Path, request, cancellationToken);

    /// <summary>Updates the target URL and/or subscribed event types of a subscription.</summary>
    public Task<Response<WebhookSubscriptionResponse>> EditAsync(string accountName, string id, EditWebhookSubscriptionRequest request, CancellationToken cancellationToken = default) =>
        _transport.PutAsync<WebhookSubscriptionResponse>($"{Path}/{Enc(accountName)}/{Enc(id)}", request, cancellationToken);

    /// <summary>Permanently removes a subscription.</summary>
    public Task<Response> RemoveAsync(string accountName, string id, CancellationToken cancellationToken = default) =>
        _transport.DeleteAsync($"{Path}/{Enc(accountName)}/{Enc(id)}", cancellationToken);

    /// <summary>Fires a synthetic test delivery. Resolves once accepted (HTTP 202).</summary>
    public Task<Response> TestAsync(string accountName, string id, CancellationToken cancellationToken = default) =>
        _transport.PostAsync($"{Path}/{Enc(accountName)}/{Enc(id)}/test", null, cancellationToken);
}

File: packages/public-api/src/WebhookService.ts
import Fetch from "./Fetch.js";
import { handleResponse } from "./handleResponse.js";
import {
  Response,
  Problem,
  CreateWebhookSubscriptionRequest,
  EditWebhookSubscriptionRequest,
  WebhookSubscriptionResponse,
  WebhookSubscriptionWithSecretResponse,
  WebhookEventGroupResponse,
} from "./Types.js";

const enc = encodeURIComponent;

// Webhook subscription management: list the event catalog, create/edit/remove a subscription, and fire
// a test delivery. Secret view/rotate, delivery-log reads and resume stay console-plane.
export default class WebhookService {
  private readonly path = `webhook`;
  constructor(private readonly fetch: Fetch) {}

  /** Lists the catalog of subscribable event types, grouped for display. */
  async listEvents(): Promise<Response<WebhookEventGroupResponse[], Problem>> {
    const response = await this.fetch.get(`${this.path}/events`);
    return handleResponse(response, async () => (await response.json()) as WebhookEventGroupResponse[]);
  }

  /**
   * Creates a webhook subscription. The response includes the signing secret exactly once — store it
   * now; it cannot be retrieved again from the data plane.
   */
  async create(request: CreateWebhookSubscriptionRequest): Promise<Response<WebhookSubscriptionWithSecretResponse, Problem>> {
    const response = await this.fetch.post(this.path, request);
    return handleResponse(response, async () => (await response.json()) as WebhookSubscriptionWithSecretResponse);
  }

  /** Updates the target URL and/or subscribed event types of a subscription. */
  async edit(account_name: string, id: string, request: EditWebhookSubscriptionRequest): Promise<Response<WebhookSubscriptionResponse, Problem>> {
    const response = await this.fetch.put(`${this.path}/${enc(account_name)}/${enc(id)}`, request);
    return handleResponse(response, async () => (await response.json()) as WebhookSubscriptionResponse);
  }

  /** Permanently removes a subscription. */
  async remove(account_name: string, id: string): Promise<Response<void, Problem>> {
    const response = await this.fetch.delete(`${this.path}/${enc(account_name)}/${enc(id)}`);
    return handleResponse(response, () => undefined);
  }

  /** Fires a synthetic test delivery. Resolves once accepted (HTTP 202). */
  async test(account_name: string, id: string): Promise<Response<void, Problem>> {
    const response = await this.fetch.post(`${this.path}/${enc(account_name)}/${enc(id)}/test`, {});
    return handleResponse(response, () => undefined);
  }
}
